File: java/src/main/java/com/inariwatch/capture/Dsn.java
package com.inariwatch.capture;

/**
 * Parsed DSN. Mirrors the Node SDK's transport.ts: secret extracted from
 * the user-info component, path normalised so {@code /capture/ID} becomes
 * {@code /api/webhooks/capture/ID}, and localhost DSNs are flagged so the
 * caller can skip TLS + HMAC.
 */
public final class Dsn {
    public final String url;
    public final String secret;
    public final String projectId;
    public final boolean isLocal;

    private Dsn(String url, String secret, String projectId, boolean isLocal) {
        this.url = url;
        this.secret = secret;
        this.projectId = projectId;
        this.isLocal = isLocal;
    }

    public static Dsn parse(String dsn) {
        String s = dsn.trim();
        boolean https;
        if (s.startsWith("https://")) {
            https = true;
            s = s.substring(8);
        } else if (s.startsWith("http://")) {
            https = false;
            s = s.substring(7);
        } else {
            throw new IllegalArgumentException("invalid DSN url: " + dsn);
        }

        int at = s.indexOf('@');
        if (at < 0) {
            throw new IllegalArgumentException("DSN missing secret");
        }
        String secret = s.substring(0, at);
        if (secret.isEmpty()) {
            throw new IllegalArgumentException("DSN missing secret");
        }
        String hostPath = s.substring(at + 1);

        int slash = hostPath.indexOf('/');
        String host = slash < 0 ? hostPath : hostPath.substring(0, slash);
        String path = slash < 0 ? "/" : hostPath.substring(slash);

        int colon = host.indexOf(':');
        String hostOnly = colon < 0 ? host : host.substring(0, colon);
        boolean isLocal = hostOnly.equals("localhost") || hostOnly.equals("127.0.0.1");

        if (!https && !isLocal) {
            throw new IllegalArgumentException("DSN must use HTTPS unless host is localhost");
        }

        String projectId;
        if (path.startsWith("/capture/")) {
            projectId = path.substring("/capture/".length()).replaceAll("/+$", "");
            path = "/api/webhooks/capture/" + projectId;
        } else if (path.startsWith("/api/webhooks/capture/")) {
            projectId = path.substring("/api/webhooks/capture/".length()).replaceAll("/+$", "");
        } else {
            throw new IllegalArgumentException("invalid DSN path: " + path);
        }
        if (projectId.isEmpty()) {
            throw new IllegalArgumentException("DSN missing project id");
        }

        String url = (https ? "https://" : "http://") + host + path;
        return new Dsn(url, secret, projectId, isLocal);
    }
}

File: csharp/src/Dsn.cs
namespace InariWatch.Capture;

/// <summary>
/// Parsed DSN. Identical wire-format to the Node SDK: secret extracted
/// from the user-info, path normalised so {/capture/ID} -> {/api/webhooks/capture/ID},
/// and localhost DSNs are flagged so callers can skip TLS + HMAC.
/// </summary>
public sealed record Dsn(string Url, string Secret, string ProjectId, bool IsLocal)
{
    public static Dsn Parse(string dsn)
    {
        string s = dsn.Trim();
        bool https;
        if (s.StartsWith("https://", StringComparison.Ordinal))
        {
            https = true;
            s = s[8..];
        }
        else if (s.StartsWith("http://", StringComparison.Ordinal))
        {
            https = false;
            s = s[7..];
        }
        else
        {
            throw new ArgumentException($"invalid DSN url: {dsn}");
        }

        int at = s.IndexOf('@');
        if (at < 0) throw new ArgumentException("DSN missing secret");
        string secret = s[..at];
        if (secret.Length == 0) throw new ArgumentException("DSN missing secret");
        string hostPath = s[(at + 1)..];

        int slash = hostPath.IndexOf('/');
        string host = slash < 0 ? hostPath : hostPath[..slash];
        string path = slash < 0 ? "/" : hostPath[slash..];

        int colon = host.IndexOf(':');
        string hostOnly = colon < 0 ? host : host[..colon];
        bool isLocal = hostOnly == "localhost" || hostOnly == "127.0.0.1";

        if (!https && !isLocal)
            throw new ArgumentException("DSN must use HTTPS unless host is localhost");

        string projectId;
        if (path.StartsWith("/capture/"))
        {
            projectId = path["/capture/".Length..].TrimEnd('/');
            path = "/api/webhooks/capture/" + projectId;
        }
        else if (path.StartsWith("/api/webhooks/capture/"))
        {
            projectId = path["/api/webhooks/capture/".Length..].TrimEnd('/');
        }
        else
        {
            throw new ArgumentException($"invalid DSN path: {path}");
        }
        if (projectId.Length == 0) throw new ArgumentException("DSN missing project id");

        string url = (https ? "https://" : "http://") + host + path;
        return new Dsn(url, secret, projectId, isLocal);
    }
}
